adcc/ExcitedStates.py

```diff
@@ -586,3 +654,60 @@ def describe_amplitudes(self, tolerance=0.01, index_format=None):
                 ret += "\n"
                 ret += separator
         return ret[:-1]
+
+    @property
+    def excitations(self):
+        """
+        Provides a list of Excitations, i.e., a view to all individual
+        excited states and their properties. Still under heavy development.
+        """
+        class Excitation:
+            pass
+
+        excitations = []
+        for i in range(self.size):
+            e = Excitation()
+            e.idx = i
+            e.__parent_state = self
+            for key in dir(self):
+                if key == "excitations":
+                    continue
+                if "__" in key or key.startswith("_"):
+                    continue  # skip "private" fields
+                if hasattr(type(self), key):
+                    if not isinstance(getattr(type(self), key), property):
+                        continue
+                else:
+                    continue
+
+                fget = getattr(type(self), key).fget
+                if not hasattr(fget, "__excitation_property"):
+                    # print(dir(fget), key, "not found.")
+                    continue
+                # Extract the kwargs passed to mark_excitation_property
+                kwargs = getattr(fget, "__excitation_property").copy()
+
+                def payload(self, key=key, kwargs=kwargs):
+                    return getattr(self.__parent_state, key)[self.idx]
+
+                setattr(type(e), kwargs["name"], property(payload))
+
+                transform = kwargs.pop("transform_to_ao", False)
+                if transform:
+                    def payload_transform(self, key=key):
+                        return transform_to_ao(
+                            getattr(self.__parent_state, key)[self.idx])
+                    setattr(type(e), kwargs["name"] + "_ao",
+                            property(payload_transform))
+            excitations.append(e)
+
+        return excitations
+
+
+def transform_to_ao(dm):
+    if not isinstance(dm, OneParticleOperator):
+        raise TypeError("Density matrix must be of type"
+                        " OneParticleOperator.")
+    density = dm.to_ao_basis()
+    density_ao = density[0] + density[1]
+    return density_ao.to_ndarray()
```

Comment:
Once the lazy stuff is merged you can just do `sum(dm.to_ao_basis())`. Also I am not sure `to_ndarray()` should be done by default. In either way this will only be one line, so I would not even have this function.

Reply:
Done. Left a comment for the lazy stuff. About the `to_ndarray()`, I'm also not sure at the moment.